False positive of fuzz

# What is the problem?

Still got false positive of fuzz on a project containing many regular expressions.

```
  9:7  error  Found a ReDoS vulnerable RegExp (2nd degree polynomial (fuzz))  redos/no-vulnerable

✖ 1 problem (1 error, 0 warnings)


webpack 5.73.0 compiled with 1 error in 213135 ms
```

That error appears in different files each time. I've set attackTimeout to null (#447) but the error doesn't disappear. Any other workarounds?

With the next parameters, recheck almost always seems to return the next errors. Those reasons are sometimes changed to timeout.

```
    "redos/no-vulnerable": [
      "error",
      {
        "ignoreErrors": false,
        "maxPatternSize": 3000,
        "maxRepeatCount": 256,
        "maxSimpleRepeatCount": 256,
        "attackTimeout": null,
        "incubationTimeout": null,
        "recallTimeout": null,
        "seedingTimeout": null,
        "timeout": 30000
      }
    ]
```

```
ERROR in
...\securemark\src\parser\header.ts
  17:17  error  Error on ReDoS vulnerablity check: timeout  redos/no-vulnerable

...\securemark\src\parser\inline\autolink\account.ts
  14:13  error  Found a ReDoS vulnerable RegExp (3rd degree polynomial (fuzz))  redos/no-vulnerable

...\securemark\src\parser\inline\autolink\hashtag.ts
  17:13  error  Found a ReDoS vulnerable RegExp (3rd degree polynomial (fuzz))  redos/no-vulnerable

✖ 3 problems (3 errors, 0 warnings)
```

--- a/modules/recheck-codec/shared/src/main/scala/codes/quine/labs/recheck/codec/package.scala
+++ b/modules/recheck-codec/shared/src/main/scala/codes/quine/labs/recheck/codec/package.scala
@@ -4,6 +4,8 @@ import scala.concurrent.duration.Duration
 import scala.concurrent.duration.MILLISECONDS
 
 import io.circe.Decoder
+import io.circe.DecodingFailure
+import io.circe.DecodingFailure.Reason.MissingField
 import io.circe.Encoder
 import io.circe.HCursor
 import io.circe.Json
@@ -109,38 +111,47 @@ package object codec {
   implicit def encodeUString: Encoder[UString] = _.asString.asJson
 
   /** A `Decoder` for `Parameters`. */
-  implicit def decodeParameters(implicit decodeLogger: Decoder[Context.Logger]): Decoder[Parameters] = (c: HCursor) =>
+  implicit def decodeParameters(implicit decodeLogger: Decoder[Context.Logger]): Decoder[Parameters] = (c: HCursor) => {
+
+    /** Returns a decoded result if `key` is found, or returns the given fallback value as a result if key is missing.
+      *
+      * It is almost similar to `HCursor#getOrElse`. However, it only falls back on missing key (not on `null` case).
+      */
+    def getOrElse[A: Decoder](key: String)(fallback: A): Decoder.Result[A] =
+      c.get[A](key) match {
+        case Left(failure) if failure.reason == MissingField => Right(fallback)
+        case result                                          => result
+      }
+
     for {
-      accelerationMode <- c.get[AccelerationMode]("accelerationMode").orElse(Right(Parameters.DefaultAccelerationMode))
-      attackLimit <- c.get[Int]("attackLimit").orElse(Right(Parameters.DefaultAttackLimit))
-      attackTimeout <- c.get[Duration]("attackTimeout").orElse(Right(Parameters.DefaultAttackTimeout))
-      checker <- c.get[Checker]("checker").orElse(Right(Parameters.DefaultChecker))
-      crossoverSize <- c.get[Int]("crossoverSize").orElse(Right(Parameters.DefaultCrossoverSize))
-      heatRatio <- c.get[Double]("heatRatio").orElse(Right(Parameters.DefaultHeatRatio))
-      incubationLimit <- c.get[Int]("incubationLimit").orElse(Right(Parameters.DefaultIncubationLimit))
-      incubationTimeout <- c.get[Duration]("incubationTimeout").orElse(Right(Parameters.DefaultIncubationTimeout))
-      logger <- c.get[Option[Context.Logger]]("logger").orElse(Right(Parameters.DefaultLogger))
-      maxAttackStringSize <- c.get[Int]("maxAttackStringSize").orElse(Right(Parameters.DefaultMaxAttackStringSize))
-      maxDegree <- c.get[Int]("maxDegree").orElse(Right(Parameters.DefaultMaxDegree))
-      maxGeneStringSize <- c.get[Int]("maxGeneStringSize").orElse(Right(Parameters.DefaultMaxGeneStringSize))
-      maxGenerationSize <- c.get[Int]("maxGenerationSize").orElse(Right(Parameters.DefaultMaxGenerationSize))
-      maxInitialGenerationSize <- c
-        .get[Int]("maxInitialGenerationSize")
-        .orElse(Right(Parameters.DefaultMaxInitialGenerationSize))
-      maxIteration <- c.get[Int]("maxIteration").orElse(Right(Parameters.DefaultMaxIteration))
-      maxNFASize <- c.get[Int]("maxNFASize").orElse(Right(Parameters.DefaultMaxNFASize))
-      maxPatternSize <- c.get[Int]("maxPatternSize").orElse(Right(Parameters.DefaultMaxPatternSize))
-      maxRecallStringSize <- c.get[Int]("maxRecallStringSize").orElse(Right(Parameters.DefaultMaxRecallStringSize))
-      maxRepeatCount <- c.get[Int]("maxRepeatCount").orElse(Right(Parameters.DefaultMaxRepeatCount))
-      maxSimpleRepeatCount <- c.get[Int]("maxSimpleRepeatCount").orElse(Right(Parameters.DefaultMaxSimpleRepeatCount))
-      mutationSize <- c.get[Int]("mutationSize").orElse(Right(Parameters.DefaultMutationSize))
-      randomSeed <- c.get[Long]("randomSeed").orElse(Right(Parameters.DefaultRandomSeed))
-      recallLimit <- c.get[Int]("recallLimit").orElse(Right(Parameters.DefaultRecallLimit))
-      recallTimeout <- c.get[Duration]("recallTimeout").orElse(Right(Parameters.DefaultRecallTimeout))
-      seeder <- c.get[Seeder]("seeder").orElse(Right(Parameters.DefaultSeeder))
-      seedingLimit <- c.get[Int]("seedingLimit").orElse(Right(Parameters.DefaultSeedingLimit))
-      seedingTimeout <- c.get[Duration]("seedingTimeout").orElse(Right(Parameters.DefaultSeedingTimeout))
-      timeout <- c.get[Duration]("timeout").orElse(Right(Parameters.DefaultTimeout))
+      accelerationMode <- getOrElse[AccelerationMode]("accelerationMode")(Parameters.DefaultAccelerationMode)
+      attackLimit <- getOrElse[Int]("attackLimit")(Parameters.DefaultAttackLimit)
+      attackTimeout <- getOrElse[Duration]("attackTimeout")(Parameters.DefaultAttackTimeout)
+      checker <- getOrElse[Checker]("checker")(Parameters.DefaultChecker)
+      crossoverSize <- getOrElse[Int]("crossoverSize")(Parameters.DefaultCrossoverSize)
+      heatRatio <- getOrElse[Double]("heatRatio")(Parameters.DefaultHeatRatio)
+      incubationLimit <- getOrElse[Int]("incubationLimit")(Parameters.DefaultIncubationLimit)
+      incubationTimeout <- getOrElse[Duration]("incubationTimeout")(Parameters.DefaultIncubationTimeout)
+      logger <- getOrElse[Option[Context.Logger]]("logger")(Parameters.DefaultLogger)
+      maxAttackStringSize <- getOrElse[Int]("maxAttackStringSize")(Parameters.DefaultMaxAttackStringSize)
+      maxDegree <- getOrElse[Int]("maxDegree")(Parameters.DefaultMaxDegree)
+      maxGeneStringSize <- getOrElse[Int]("maxGeneStringSize")(Parameters.DefaultMaxGeneStringSize)
+      maxGenerationSize <- getOrElse[Int]("maxGenerationSize")(Parameters.DefaultMaxGenerationSize)
+      maxInitialGenerationSize <- getOrElse[Int]("maxInitialGenerationSize")(Parameters.DefaultMaxInitialGenerationSize)
+      maxIteration <- getOrElse[Int]("maxIteration")(Parameters.DefaultMaxIteration)
+      maxNFASize <- getOrElse[Int]("maxNFASize")(Parameters.DefaultMaxNFASize)
+      maxPatternSize <- getOrElse[Int]("maxPatternSize")(Parameters.DefaultMaxPatternSize)
+      maxRecallStringSize <- getOrElse[Int]("maxRecallStringSize")(Parameters.DefaultMaxRecallStringSize)
+      maxRepeatCount <- getOrElse[Int]("maxRepeatCount")(Parameters.DefaultMaxRepeatCount)
+      maxSimpleRepeatCount <- getOrElse[Int]("maxSimpleRepeatCount")(Parameters.DefaultMaxSimpleRepeatCount)
+      mutationSize <- getOrElse[Int]("mutationSize")(Parameters.DefaultMutationSize)
+      randomSeed <- getOrElse[Long]("randomSeed")(Parameters.DefaultRandomSeed)
+      recallLimit <- getOrElse[Int]("recallLimit")(Parameters.DefaultRecallLimit)
+      recallTimeout <- getOrElse[Duration]("recallTimeout")(Parameters.DefaultRecallTimeout)
+      seeder <- getOrElse[Seeder]("seeder")(Parameters.DefaultSeeder)
+      seedingLimit <- getOrElse[Int]("seedingLimit")(Parameters.DefaultSeedingLimit)
+      seedingTimeout <- getOrElse[Duration]("seedingTimeout")(Parameters.DefaultSeedingTimeout)
+      timeout <- getOrElse[Duration]("timeout")(Parameters.DefaultTimeout)
     } yield Parameters(
       accelerationMode,
       attackLimit,
@@ -171,13 +182,16 @@ package object codec {
       seedingTimeout,
       timeout
     )
+  }
 
   /** A `Decoder` for `Duration`. */
-  implicit def decodeDuration: Decoder[Duration] = {
-    import io.circe.disjunctionCodecs._
-
-    Decoder[Either[Unit, Int]].map(_.fold(_ => Duration.Inf, Duration(_, MILLISECONDS)))
-  }
+  implicit def decodeDuration: Decoder[Duration] = (c: HCursor) =>
+    if (c.value.isNull) Right(Duration.Inf)
+    else if (c.value.isNumber) c.value.asNumber.flatMap(_.toInt) match {
+      case Some(d) => Right(Duration(d, MILLISECONDS))
+      case None    => Left(DecodingFailure("Duration", c.history))
+    }
+    else Left(DecodingFailure("Duration", c.history))
 
   /** A `Decoder` for `Checker`. */
   implicit def decodeChecker: Decoder[Checker] =

--- a/modules/recheck-codec/shared/src/main/scala/codes/quine/labs/recheck/codec/package.scala
+++ b/modules/recheck-codec/shared/src/main/scala/codes/quine/labs/recheck/codec/package.scala
@@ -111,36 +111,36 @@ package object codec {
   /** A `Decoder` for `Parameters`. */
   implicit def decodeParameters(implicit decodeLogger: Decoder[Context.Logger]): Decoder[Parameters] = (c: HCursor) =>
     for {
-      accelerationMode <- c.getOrElse[AccelerationMode]("accelerationMode")(Parameters.DefaultAccelerationMode)
-      attackLimit <- c.getOrElse[Int]("attackLimit")(Parameters.DefaultAttackLimit)
-      attackTimeout <- c.getOrElse[Duration]("attackTimeout")(Parameters.DefaultAttackTimeout)
-      checker <- c.getOrElse[Checker]("checker")(Parameters.DefaultChecker)
-      crossoverSize <- c.getOrElse[Int]("crossoverSize")(Parameters.DefaultCrossoverSize)
-      heatRatio <- c.getOrElse[Double]("heatRatio")(Parameters.DefaultHeatRatio)
-      incubationLimit <- c.getOrElse[Int]("incubationLimit")(Parameters.DefaultIncubationLimit)
-      incubationTimeout <- c.getOrElse[Duration]("incubationTimeout")(Parameters.DefaultIncubationTimeout)
-      logger <- c.getOrElse[Option[Context.Logger]]("logger")(Parameters.DefaultLogger)
-      maxAttackStringSize <- c.getOrElse[Int]("maxAttackStringSize")(Parameters.DefaultMaxAttackStringSize)
-      maxDegree <- c.getOrElse[Int]("maxDegree")(Parameters.DefaultMaxDegree)
-      maxGeneStringSize <- c.getOrElse[Int]("maxGeneStringSize")(Parameters.DefaultMaxGeneStringSize)
-      maxGenerationSize <- c.getOrElse[Int]("maxGenerationSize")(Parameters.DefaultMaxGenerationSize)
-      maxInitialGenerationSize <- c.getOrElse[Int]("maxInitialGenerationSize")(
-        Parameters.DefaultMaxInitialGenerationSize
-      )
-      maxIteration <- c.getOrElse[Int]("maxIteration")(Parameters.DefaultMaxIteration)
-      maxNFASize <- c.getOrElse[Int]("maxNFASize")(Parameters.DefaultMaxNFASize)
-      maxPatternSize <- c.getOrElse[Int]("maxPatternSize")(Parameters.DefaultMaxPatternSize)
-      maxRecallStringSize <- c.getOrElse[Int]("maxRecallStringSize")(Parameters.DefaultMaxRecallStringSize)
-      maxRepeatCount <- c.getOrElse[Int]("maxRepeatCount")(Parameters.DefaultMaxRepeatCount)
-      maxSimpleRepeatCount <- c.getOrElse[Int]("maxSimpleRepeatCount")(Parameters.DefaultMaxSimpleRepeatCount)
-      mutationSize <- c.getOrElse[Int]("mutationSize")(Parameters.DefaultMutationSize)
-      randomSeed <- c.getOrElse[Long]("randomSeed")(Parameters.DefaultRandomSeed)
-      recallLimit <- c.getOrElse[Int]("recallLimit")(Parameters.DefaultRecallLimit)
-      recallTimeout <- c.getOrElse[Duration]("recallTimeout")(Parameters.DefaultRecallTimeout)
-      seeder <- c.getOrElse[Seeder]("seeder")(Parameters.DefaultSeeder)
-      seedingLimit <- c.getOrElse[Int]("seedingLimit")(Parameters.DefaultSeedingLimit)
-      seedingTimeout <- c.getOrElse[Duration]("seedingTimeout")(Parameters.DefaultSeedingTimeout)
-      timeout <- c.getOrElse[Duration]("timeout")(Parameters.DefaultTimeout)
+      accelerationMode <- c.get[AccelerationMode]("accelerationMode").orElse(Right(Parameters.DefaultAccelerationMode))
+      attackLimit <- c.get[Int]("attackLimit").orElse(Right(Parameters.DefaultAttackLimit))
+      attackTimeout <- c.get[Duration]("attackTimeout").orElse(Right(Parameters.DefaultAttackTimeout))
+      checker <- c.get[Checker]("checker").orElse(Right(Parameters.DefaultChecker))
+      crossoverSize <- c.get[Int]("crossoverSize").orElse(Right(Parameters.DefaultCrossoverSize))
+      heatRatio <- c.get[Double]("heatRatio").orElse(Right(Parameters.DefaultHeatRatio))
+      incubationLimit <- c.get[Int]("incubationLimit").orElse(Right(Parameters.DefaultIncubationLimit))
+      incubationTimeout <- c.get[Duration]("incubationTimeout").orElse(Right(Parameters.DefaultIncubationTimeout))
+      logger <- c.get[Option[Context.Logger]]("logger").orElse(Right(Parameters.DefaultLogger))
+      maxAttackStringSize <- c.get[Int]("maxAttackStringSize").orElse(Right(Parameters.DefaultMaxAttackStringSize))
+      maxDegree <- c.get[Int]("maxDegree").orElse(Right(Parameters.DefaultMaxDegree))
+      maxGeneStringSize <- c.get[Int]("maxGeneStringSize").orElse(Right(Parameters.DefaultMaxGeneStringSize))
+      maxGenerationSize <- c.get[Int]("maxGenerationSize").orElse(Right(Parameters.DefaultMaxGenerationSize))
+      maxInitialGenerationSize <- c
+        .get[Int]("maxInitialGenerationSize")
+        .orElse(Right(Parameters.DefaultMaxInitialGenerationSize))
+      maxIteration <- c.get[Int]("maxIteration").orElse(Right(Parameters.DefaultMaxIteration))
+      maxNFASize <- c.get[Int]("maxNFASize").orElse(Right(Parameters.DefaultMaxNFASize))
+      maxPatternSize <- c.get[Int]("maxPatternSize").orElse(Right(Parameters.DefaultMaxPatternSize))
+      maxRecallStringSize <- c.get[Int]("maxRecallStringSize").orElse(Right(Parameters.DefaultMaxRecallStringSize))
+      maxRepeatCount <- c.get[Int]("maxRepeatCount").orElse(Right(Parameters.DefaultMaxRepeatCount))
+      maxSimpleRepeatCount <- c.get[Int]("maxSimpleRepeatCount").orElse(Right(Parameters.DefaultMaxSimpleRepeatCount))
+      mutationSize <- c.get[Int]("mutationSize").orElse(Right(Parameters.DefaultMutationSize))
+      randomSeed <- c.get[Long]("randomSeed").orElse(Right(Parameters.DefaultRandomSeed))
+      recallLimit <- c.get[Int]("recallLimit").orElse(Right(Parameters.DefaultRecallLimit))
+      recallTimeout <- c.get[Duration]("recallTimeout").orElse(Right(Parameters.DefaultRecallTimeout))
+      seeder <- c.get[Seeder]("seeder").orElse(Right(Parameters.DefaultSeeder))
+      seedingLimit <- c.get[Int]("seedingLimit").orElse(Right(Parameters.DefaultSeedingLimit))
+      seedingTimeout <- c.get[Duration]("seedingTimeout").orElse(Right(Parameters.DefaultSeedingTimeout))
+      timeout <- c.get[Duration]("timeout").orElse(Right(Parameters.DefaultTimeout))
     } yield Parameters(
       accelerationMode,
       attackLimit,
@@ -173,9 +173,10 @@ package object codec {
     )
 
   /** A `Decoder` for `Duration`. */
-  implicit def decodeDuration: Decoder[Duration] = Decoder[Option[Int]].map {
-    case Some(d) => Duration(d, MILLISECONDS)
-    case None    => Duration.Inf
+  implicit def decodeDuration: Decoder[Duration] = {
+    import io.circe.disjunctionCodecs._
+
+    Decoder[Either[Unit, Int]].map(_.fold(_ => Duration.Inf, Duration(_, MILLISECONDS)))
   }
 
   /** A `Decoder` for `Checker`. */

--- a/modules/recheck-codec/shared/src/test/scala/codes/quine/labs/recheck/codec/CodecSuite.scala
+++ b/modules/recheck-codec/shared/src/test/scala/codes/quine/labs/recheck/codec/CodecSuite.scala
@@ -215,6 +215,8 @@ class CodecSuite extends munit.FunSuite {
   test("codec.decodeDuration") {
     assertEquals(decodeDuration.decodeJson(Json.Null), Right(Duration.Inf))
     assertEquals(decodeDuration.decodeJson(100.asJson), Right(Duration(100, MILLISECONDS)))
+    assertEquals(decodeDuration.decodeJson(Json.True), Left(DecodingFailure("Duration", List.empty)))
+    assertEquals(decodeDuration.decodeJson(123.456.asJson), Left(DecodingFailure("Duration", List.empty)))
   }
 
   test("codec.decodeChecker") {

--- a/modules/recheck-codec/shared/src/test/scala/codes/quine/labs/recheck/codec/CodecSuite.scala
+++ b/modules/recheck-codec/shared/src/test/scala/codes/quine/labs/recheck/codec/CodecSuite.scala
@@ -135,6 +135,24 @@ class CodecSuite extends munit.FunSuite {
     assertEquals(
       decodeParameters.decodeJson(
         Json.obj(
+          "attackTimeout" := Json.Null,
+          "incubationTimeout" := Json.Null,
+          "recallTimeout" := Json.Null,
+          "seedingTimeout" := Json.Null
+        )
+      ),
+      Right(
+        Parameters(
+          attackTimeout = Duration.Inf,
+          incubationTimeout = Duration.Inf,
+          recallTimeout = Duration.Inf,
+          seedingTimeout = Duration.Inf
+        )
+      )
+    )
+    assertEquals(
+      decodeParameters.decodeJson(
+        Json.obj(
           "checker" := "fuzz",
           "timeout" := 123,
           "logger" := Json.arr(),

